pypeit/core/datacube.py

```diff
@@ -0,0 +1,338 @@
+            For example, a 0 would indicate that a pixel originates from the first spec2d
+            frame listed in the input file. a 1 would indicate that this pixel originates
+            from the second spec2d file, and so forth.
+        dspat (float):
+            The size of each spaxel on the sky (in degrees)
+        dwv (float):
+            The size of each wavelength pixel (in Angstroms)
+        numfiles (int, optional):
+            Number of spec2d files included. If not provided, it will be calculated from all_idx
+
+    Returns:
+        `numpy.ndarray`_ : A 1D numpy array the same size as all_sci, containing
+        relative wavelength dependent weights for each input pixel.
+    """
+    msgs.info("Calculating the optimal weights of each pixel")
+    # Determine number of files
+    if numfiles is None:
+        numfiles = np.unique(all_idx).size
+
+    # Generate a white light image of *all* data
+    msgs.info("Generating global white light image")
+    whitelight_Img, ivar = generate_whiteLightImgs(all_ra, all_dec, all_wave, all_sci, all_wghts, np.zeros(all_ra.size),
+                                                dspat, numfiles=1)
+    whitelight_Img = whitelight_Img[:, :, 0]
+    # Find the location of the object with the highest S/N in the combined white light image
+    idx_max = np.unravel_index(np.argmax(whitelight_Img), whitelight_Img.shape)
+    msgs.info("Highest S/N object located at spaxel (x, y) = {0:d}, {1:d}".format(idx_max[0], idx_max[1]))
+
+    # Generate a master 2D WCS to register all frames
+    coord_min = [np.min(all_ra), np.min(all_dec), np.min(all_wave)]
+    coord_dlt = [dspat, dspat, dwv]
+    whitelightWCS = generate_masterWCS(coord_min, coord_dlt)
+    # Make the bin edges to be at +/- 1 pixels around the maximum (i.e. summing 9 pixels total)
+    numwav = int((np.max(all_wave) - np.min(all_wave)) / dwv)
+    xbins = np.array([idx_max[0]-1, idx_max[0]+2]) - 0.5
+    ybins = np.array([idx_max[1]-1, idx_max[1]+2]) - 0.5
+    spec_bins = np.arange(1 + numwav) - 0.5
+    bins = (xbins, ybins, spec_bins)
+
+    # Extract the spectrum of the highest S/N object
+    all_spec = np.zeros((numwav, numfiles))
+    all_snr = np.zeros((numwav, numfiles))
+    for ff in range(numfiles):
+        msgs.info("Extracting spectrum of highest S/N detection from frame {0:d}/{1:d}".format(ff + 1, numfiles))
+        ww = (all_idx == ff)
+        # Extract the spectrum
+        pix_coord = whitelightWCS.wcs_world2pix(np.vstack((all_ra[ww], all_dec[ww], all_wave[ww] * 1.0E-10)).T, 0)
+        spec, edges = np.histogramdd(pix_coord, bins=bins, weights=all_sci[ww])
+        norm, edges = np.histogramdd(pix_coord, bins=bins)
+        ivar, edges = np.histogramdd(pix_coord, bins=bins, weights=all_ivar[ww])
+        nrmSpec = (norm > 0) / (norm + (norm == 0))
+        spec = (spec*nrmSpec)[0, 0, :]
+        all_spec[:, ff] = spec.copy()
+        all_snr[:, ff] = spec*np.sqrt(ivar[0, 0, :])
+
```

Comment:
@rcooke-ast take a look at core.coadd.sn_weights. I think that does something very similar to what you are doing here. That code is very well tested, and ideally we would not be doing this in multiple ways.

Reply:
Let's talk about this on Slack